R/class-Node.R

```diff
@@ -0,0 +1,88 @@
+check_node <- function(object) {
+
+  errors <- character()
+
+  # Node color must be a string or number
+  if (!is.null(object@color) & !is.character(object@color) & !is.numeric(object@color)) {
+    errors <- c(errors, "Node color must be a string or number")
+  }
+
+  # Node weight must be a number
+  if (!is.null(object@weight) & !is.numeric(object@weight)) {
+    errors <- c(errors, "Node weight must be a number")
+  }
+
+  return(if (length(errors) == 0) TRUE else errors)
+}
+
+
+#' Node
+#' 
+#' A class for representing nodes in a network
+#' 
+#' @slot id string a unique identifier for the node
+#' @slot color string or numeric that determines the color of the node. Optional.
+#' @slot weight numeric value associated with the node, such as timestamp or other node-associated data. Optional.
+#' 
+#' @name Node-class
+#' @rdname Node-class
+#' @export
+Node <- setClass("Node", 
+  representation(
+    id = "character",
+    color = "ANY",
+    weight = "ANY"
```

Comment:
Remind me what's weight here? Truly anything? Do we need a general metadata slot? Should it take a named list maybe?

Reply:
Weight needs to be numeric (if there's not a check for that i'll go add it), but generally nodes don't have weights so i like having the default of NULL. Links have a default weight because we think of even unweighted networks as having 0s (no link) or 1s (a link) between node pairs.
I suppose setting the weight to _NA_numeric would be cleaner...

Reply:
Still lost sry. How is the network visually impacted by this weight?

Reply:
I'm pretty sure we need a way for these networks to carry arbitrary data/annotations about nodes and maybe links too

Reply:
> How is the network visually impacted by this weight?

Weight is the data associated with the node or edge that we should use when calculating network/node/link metrics (think weighted degree, modularity, etc). So for correlation, the link weight would be the correlation coeff. Think of it as the primary numeric data describing the node or edge. Weight can, but does not have to be, used to color the node or edge. It can be used directly like if we colored the bipartite network edges by correlation coeff, or indirectly like just using the sign of the correlation coeff to color edges. 

> I'm pretty sure we need a way for these networks to carry arbitrary data/annotations about nodes and maybe links too

Yes i think you're right, which is why i have nodes and links set up this way as their own classes. Basically the color, weight, etc is all just data about the node/link, but it's common enough in networks that we might as well just call it what it is. For the arbitrary data, we'd have a slot called something like "additionalData" that would be a named list. Additional data could be just more values for the object, like adding pvalue and adjusted pvalue to each link, or could be vectors or even data frames, like if each node is a time point and the data attached to it is the state of the system at that point in time. The latter options are less likely in the near term but totally possible. Anyways, i think that's how this framework would grow to accomodate extra data, and additionally i think that can be implemented later when we are really seeing the pvalues and such.

Reply:
I need to think about this for a min

Reply:
Maybe all we need is to add at a high level a way for the network to keep track of what varspecs map to what network attribute.